File: Programmers/pro_67256_키패드누르기.py
def calPos(num):    # {*:-1} {0:-2} {#:-3} 매핑
    if(num<=0):
        num= -num + 9
    col = num%3
    if(col ==0):
        col=3
    row = (num-1)//3
    return col*10+row

def calDist(a,b):    
    return abs(a//10-b//10) + abs(a%10-b%10)    

def solution(numbers, hand):
    leftNum = [1,4,7]
    rightNum = [3,6,9]
    
    curLeft = -1 # {*:-1}
    curRight = -3 # {#:-3}
    
    answer = ''
    for i in numbers:
        if(i==0):
            i=-2 # {0:-2}
            
        isRight = True
        if i in leftNum:
            isRight=False
        elif i in rightNum:
            isRight=True
        else:
            leftDist = calDist(calPos(curLeft),calPos(i))
            rightDist = calDist(calPos(curRight),calPos(i))
            if(leftDist<rightDist or (leftDist==rightDist and hand=='left')):
                isRight=False
        if(isRight):
            answer+='R'
            curRight=i
        else:
            answer+='L'
            curLeft=i
    return answer

File: Programmers/pro_67256_키패드 누르기.js
function solution(numbers, hand) {
    const getDist = (a, b) => {
        [a, b] = [a - 1, b - 1]
        let [r1, c1] = [parseInt(a / 3), parseInt(a % 3)]
        let [r2, c2] = [parseInt(b / 3), parseInt(b % 3)]
        return Math.abs(c2 - c1) + Math.abs(r2 - r1)
    }
    var answer = '';
    let [left, right] = [10, 12]
    for (let i = 0, len = numbers.length; i < len; i++) {
        let targetNum = numbers[i]
        if (targetNum === 0) targetNum = 11
        if (targetNum % 3 === 1) {
            answer += "L"
            left = targetNum
        } else if (targetNum % 3 === 0) {
            answer += "R"
            right = targetNum
        } else {
            let [leftDist, rightDist] = [getDist(left, targetNum), getDist(right, targetNum)]
            if (hand === "left" && leftDist === rightDist || leftDist < rightDist) {
                answer += "L"
                left = targetNum
            } else {
                answer += "R"
                right = targetNum
            }
        }
    }
    return answer;
}